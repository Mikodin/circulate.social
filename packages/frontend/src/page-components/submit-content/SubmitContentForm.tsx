import { Fragment, useState, useEffect } from 'react';
import { useRouter } from 'next/router';
import axios from 'axios';

import { ZoneId, ZonedDateTime, LocalDateTime } from '@js-joda/core';
import '@js-joda/timezone';

import { AntdMoment } from '@circulate/types';

import {
  Form,
  Input,
  Button,
  Alert,
  DatePicker,
  TimePicker,
  Select,
} from 'antd';

import css from './SubmitContentForm.module.scss';

export const SUBMIT_EVENT_ENDPOINT =
  'https://z3edrz53yg.execute-api.us-east-1.amazonaws.com/dev/events/create';

export interface Props {
  jwtToken: string;
  seedCircleId?: string;
}

export const AVAILABLE_TIMEZONES = ZoneId.getAvailableZoneIds();
const TimezoneSelects = AVAILABLE_TIMEZONES.map((timeZone) => (
  <Select.Option value={timeZone} key={timeZone}>
    {timeZone}
  </Select.Option>
));

const SubmitContentForm = (props: Props): JSX.Element => {
  const [form] = Form.useForm();
  const router = useRouter();
  const { seedCircleId } = props;

  const [isDatePickerOpen, setIsDatePickerOpen] = useState(false);
  const [isEventForm, setIsEventForm] = useState(false);
  const [
    isFetchCreateContentInFlight,
    setIsFetchCreateContentInFlight,
  ] = useState(false);

  const [isFetchCreateContentError, setIsFetchCreateContentError] = useState(
    false
  );

  const userTimeZone = ZonedDateTime.now().zone().toString();

  interface FormValues {
    title: string;
    link: string;
    date: AntdMoment & string;
    time: AntdMoment & string;
    whyShare: string;
    cost: number;
    timezone: string;
  }

  useEffect(() => {
    if (!isEventForm) {
      form.setFieldsValue({ date: '' });
      form.setFieldsValue({ time: '' });
      form.setFieldsValue({ timezone: userTimeZone });
    }
  }, [isEventForm]);

  async function onFinish(formValues: FormValues): Promise<void> {
    const { title, whyShare, date, time, timezone } = formValues;

    let dateTimeStr: undefined | ZonedDateTime;
    if (date && time) {
      const dateObject = date.toObject();
      const timeObject = time.toObject();

      const { years, months, date: dateStr } = dateObject;
      const { hours, minutes } = timeObject;
      const ldt = LocalDateTime.of(years, months, dateStr, hours, minutes);
      dateTimeStr = ZonedDateTime.of(ldt, ZoneId.of(timezone));
    }
    try {
      setIsFetchCreateContentInFlight(true);

      await axios.post(
        SUBMIT_EVENT_ENDPOINT,
        {
          name: title,
          description: whyShare,
          circleId: seedCircleId,
          dateTime: dateTimeStr && dateTimeStr.toString(),
        },
        { headers: { Authorization: props.jwtToken } }
      );

      setIsFetchCreateContentInFlight(false);
      setIsFetchCreateContentError(false);
      router.push(`/circles/${seedCircleId}`);

      return;
    } catch (e) {
      console.error(e);
      setIsFetchCreateContentInFlight(false);
      setIsFetchCreateContentError(true);
    }
  }

  return (
    <Fragment>
      <Form
        className={css.form}
        form={form}
        hideRequiredMark
        name="horizontal_login"
        size="large"
        layout="vertical"
        onFinish={onFinish}
        initialValues={
          {
            title: '',
            link: '',
            date: '',
            time: '',
            whyShare: '',
            cost: 0,
            timezone: userTimeZone,
          } as FormValues
        }
        onValuesChange={(formValues: FormValues): void => {
          setIsFetchCreateContentError(false);
          if (formValues.date) {
            setIsEventForm(true);
          }
        }}
      >
        <Form.Item
          name="title"
          rules={[{ required: true, message: 'Please give your post a Title' }]}
          label="Title of post"
        >
          <Input placeholder="Title" type="text" />
        </Form.Item>

        <Form.Item name="link" label="Link to content">
          <Input placeholder="Link" type="text" />
        </Form.Item>

        <Form.Item name="date" label="Event date">
          <DatePicker
            use12Hours={true}
            open={isDatePickerOpen}
            onOpenChange={(isOpen): void => {
              setIsDatePickerOpen(isOpen);
              if (!isOpen && !form.getFieldValue('date')) {
                setIsEventForm(false);
              } else if (form.getFieldValue('date')) {
                setIsEventForm(true);
              }
            }}
            renderExtraFooter={(): JSX.Element => (
              <Button
                onClick={(): void => {
                  setIsDatePickerOpen(false);
                  setIsEventForm(false);
                }}
              >
                Clear
              </Button>
            )}
            allowClear={false}
          />
        </Form.Item>

        {isEventForm && (
          <>
            <Form.Item name="time" label="Event time">
              <TimePicker showSecond={false} use12Hours={true} />
            </Form.Item>

            <Form.Item name="timezone" label="Timezone">
              <Select
                showSearch
                style={{ width: 200 }}
                placeholder="Timezone"
                optionFilterProp="children"
                filterOption={(input, option) =>
                  option &&
                  option.children &&
                  option.children.toLowerCase().indexOf(input.toLowerCase()) >=
                    0
                }
              >
                <Select.OptGroup>Americas</Select.OptGroup>
                {TimezoneSelects}
              </Select>
            </Form.Item>

            <Form.Item name="cost" label="Event cost">
              <Input placeholder="Cost" type="text" />
            </Form.Item>
          </>
        )}
        <Form.Item name="whyShare" label="Why are you sharing this?">
          <Input.TextArea
            rows={3}
            placeholder="Why are you sharing this?"
            // onFocus={(): void => setShowDescriptionPopover(true)}
            style={{ fontSize: '16px' }}
          />
        </Form.Item>

        {isFetchCreateContentError && (
          <Alert
            message="Something went wrong, please try again."
            type="error"
            showIcon
            closable
            banner
            style={{ marginBottom: '10px' }}
          />
        )}
        <Form.Item shouldUpdate={true}>
          {(): JSX.Element => {
            const isTitleTouched = form.isFieldTouched('title');
            const isFormTouched = isTitleTouched;
            return (
              <Button
                data-testid="submitButton"
                loading={isFetchCreateContentInFlight}
                type="primary"
                htmlType="submit"
                size="large"
                block
                disabled={Boolean(
                  !isFormTouched ||
                    form.getFieldsError().filter(({ errors }) => errors.length)
                      .length
                )}
              >
                {isFetchCreateContentInFlight ? 'Submitting...' : 'Submit'}
              </Button>
            );
          }}
        </Form.Item>
      </Form>
    </Fragment>
  );
};

export default SubmitContentForm;
