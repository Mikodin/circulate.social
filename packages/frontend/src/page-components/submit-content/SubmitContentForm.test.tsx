import {
  render,
  RenderResult,
  act,
  fireEvent,
  waitFor,
} from '@testing-library/react';
import axios from 'axios';
import { ZonedDateTime } from '@js-joda/core';

import SubmitContentForm, {
  SUBMIT_CONTENT_ENDPOINT,
  Props,
} from './SubmitContentForm';

jest.mock('axios');
const mockedAxios = axios as jest.Mocked<typeof axios>;
mockedAxios.post.mockImplementation(() => Promise.resolve(true));

const mockRouterPushSpy = jest.fn(() => true);
jest.mock('next/router', () => ({
  // eslint-disable-next-line
  useRouter() {
    return {
      push: mockRouterPushSpy,
      route: '/',
      pathname: '',
      query: '',
      asPath: '',
    };
  },
}));

const mockOnFormCompletion = jest.fn();

const userTimeZone = ZonedDateTime.now().zone().toString();

const defaultProps = {
  jwtToken: '123-asd',
  seedCircleId: 'asdf-fdsa',
  onFormCompletion: mockOnFormCompletion,
};

function renderContainer(props?: Props): RenderResult {
  return render(<SubmitContentForm {...defaultProps} {...props} />);
}

// eslint-disable-next-line
function getAllFields(container: RenderResult) {
  const { queryByPlaceholderText, queryByText } = container;

  const inputTitle = queryByPlaceholderText(/Title/i);
  const inputLink = queryByPlaceholderText(/Link/i);
  const selectDate = queryByPlaceholderText(/Select date/i);
  const selectTime = queryByPlaceholderText(/Select time/i);
  const selectTimezone = queryByText(userTimeZone);
  const inputWhyShare = queryByPlaceholderText(/Why are you sharing this?/i);

  const buttonSubmit = queryByPlaceholderText(/Submit/i);

  return {
    inputTitle,
    inputLink,
    selectDate,
    selectTime,
    selectTimezone,
    inputWhyShare,
    buttonSubmit,
  };
}

async function selectADateFromDatePicker(
  container: RenderResult
): Promise<RenderResult> {
  const { queryByText } = container;
  const { selectDate } = getAllFields(container);
  await act(async () => {
    await fireEvent.mouseDown(selectDate);
    await fireEvent.click(queryByText(/15/i));
  });

  return container;
}

async function selectATimeFromTimePicker(
  container: RenderResult
): Promise<RenderResult> {
  const { queryAllByText, queryByText } = container;
  const { selectTime } = getAllFields(container);
  await act(async () => {
    await fireEvent.mouseDown(selectTime);
    await fireEvent.click(queryAllByText(/07/i)[0]);
    await fireEvent.click(queryByText(/Ok/i));
  });

  return container;
}

async function selectTimezoneFromTimezonePicker(
  container: RenderResult
): Promise<RenderResult> {
  const { queryAllByText } = container;
  const { selectTimezone } = getAllFields(container);
  await act(async () => {
    await fireEvent.mouseDown(selectTimezone);
  });

  await act(async () => {
    await fireEvent.click(queryAllByText(userTimeZone)[0]);
  });

  return container;
}

interface FieldToPopulate {
  element: HTMLElement;
  value: string | number;
}
async function populateForm(
  container: RenderResult,
  fieldsToPopulate: FieldToPopulate[]
): Promise<RenderResult> {
  fieldsToPopulate.forEach((field) => {
    const { element, value } = field;
    act(() => {
      fireEvent.change(element, {
        target: { value },
      });
    });
  });

  return container;
}

describe('SubmitContentForm', () => {
  it('Should render', () => {
    const container = renderContainer();
    expect(container).toMatchSnapshot();
  });

  describe('When there is no value in the date input', () => {
    it('Should not render selectTimezone,or selectTime', () => {
      const container = renderContainer();
      const { selectTimezone, selectTime } = getAllFields(container);
      expect(selectTimezone).not.toBeTruthy();
      expect(selectTime).not.toBeTruthy();
    });
  });

  describe('When a user inputs a date', () => {
    it('should render the "event" flow. selectTimezone,and selectTime should all be visible', async () => {
      const container = renderContainer();

      await selectADateFromDatePicker(container);

      const { selectTimezone, selectTime } = getAllFields(container);
      expect(selectTime).toBeTruthy();
      expect(selectTimezone).toBeTruthy();
    });

    it('should set the timezone by default to the users timezone returned by Joda', async () => {
      const container = renderContainer();
      const { queryByText } = container;

      await selectADateFromDatePicker(container);

      expect(queryByText(userTimeZone)).toBeTruthy();
    });
  });

  describe('On submit of a complete ContentForm', () => {
    beforeEach(() => {
      mockedAxios.post.mockClear();
      mockOnFormCompletion.mockClear();
    });
    const inputtedTitleValue = 'The greatest Circle ever';
    const inputtedLinkValue = 'https://circulate.social';
    // const inputtedDateValue = '2020-05-20';
    const inputtedWhyShareValue =
      "Because it's a great way to share and receive content that I care about";

    async function renderCompleteForm(
      isEventForm: boolean
    ): Promise<RenderResult> {
      const container = renderContainer();
      const fields = getAllFields(container);
      const fieldsToPopulate = [
        {
          element: fields.inputTitle,
          value: inputtedTitleValue,
        },
        {
          element: fields.inputLink,
          value: inputtedLinkValue,
        },
        {
          element: fields.selectDate,
          value: '15',
        },
        {
          element: fields.inputWhyShare,
          value: inputtedWhyShareValue,
        },
      ];

      const populatedContentContainer = await populateForm(
        container,
        fieldsToPopulate
      );

      if (isEventForm) {
        const populatedEventContainer = await selectADateFromDatePicker(
          populatedContentContainer
        );
        return populatedEventContainer;
      }

      return populatedContentContainer;
    }

    it('Should fire off Axios.post on Submit and route to the Circle', async () => {
      const { queryByText } = await renderCompleteForm(false);
      const submitButton = queryByText(/Submit/i);
      mockedAxios.post.mockImplementationOnce(() => Promise.resolve(true));

      await act(async () => {
        fireEvent.submit(submitButton);
      });
      expect(mockedAxios.post).toHaveBeenCalledWith(
        SUBMIT_CONTENT_ENDPOINT,
        {
          circleId: [defaultProps.seedCircleId],
          title: inputtedTitleValue,
          description: inputtedWhyShareValue,
          link: inputtedLinkValue,
        },
        { headers: { Authorization: defaultProps.jwtToken } }
      );
      expect(mockOnFormCompletion).toHaveBeenCalledWith(inputtedTitleValue);
    });

    it('Should render a "Submitting..." loading component', async () => {
      const { queryByText } = await renderCompleteForm(false);
      const submitButton = queryByText(/Submit/i);
      act(() => {
        fireEvent.submit(submitButton);
      });
      await waitFor(() => expect(queryByText(/Submitting.../i)).toBeTruthy());
    });

    describe('on submit of an Event form', () => {
      it('Should call Axios.post on Submit and props.onFormCompletion', async () => {
        const basicContainer = await renderCompleteForm(true);
        const containerToShowTimeSelect = await selectADateFromDatePicker(
          basicContainer
        );
        const containerWithSelectedTime = await selectATimeFromTimePicker(
          containerToShowTimeSelect
        );

        const container = await selectTimezoneFromTimezonePicker(
          containerWithSelectedTime
        );
        const { queryByText } = container;

        const submitButton = queryByText(/Submit/i);
        mockedAxios.post.mockImplementationOnce(() => Promise.resolve(true));

        await act(async () => {
          fireEvent.submit(submitButton);
        });

        // @TODO - Make this adaptive to the current month
        const expectedDateTime =
          userTimeZone === 'America/Los_Angeles'
            ? '2020-10-15T07:00-07:00[America/Los_Angeles]'
            : '2020-10-15T07:00Z[UTC]';
        expect(mockedAxios.post).toHaveBeenCalledWith(
          SUBMIT_CONTENT_ENDPOINT,
          {
            circleId: [defaultProps.seedCircleId],
            title: inputtedTitleValue,
            description: inputtedWhyShareValue,
            dateTime: expectedDateTime,
            link: inputtedLinkValue,
          },
          { headers: { Authorization: defaultProps.jwtToken } }
        );

        expect(mockOnFormCompletion).toHaveBeenCalledWith(inputtedTitleValue);
      });

      describe('When time is not inputted', () => {
        it('Should not fire off Axios.post on Submit', async () => {
          const basicContainer = await renderCompleteForm(true);
          const container = await selectADateFromDatePicker(basicContainer);
          const { queryByText } = container;

          const submitButton = queryByText(/Submit/i);
          mockedAxios.post.mockImplementationOnce(() => Promise.resolve(true));

          await act(async () => {
            fireEvent.submit(submitButton);
          });

          expect(mockedAxios.post).not.toHaveBeenCalled();
          await waitFor(() =>
            expect(
              queryByText(/Since you have a date - there must be a time./i)
            ).toBeTruthy()
          );
        });
      });
    });
  });
});
