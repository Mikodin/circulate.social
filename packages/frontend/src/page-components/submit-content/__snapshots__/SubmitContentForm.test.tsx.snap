// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubmitContentForm Should render 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <form
        class="ant-form ant-form-vertical ant-form-large form"
        id="horizontal_login"
      >
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="horizontal_login_title"
              title="Title of post"
            >
              Title of post
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input ant-input-lg"
                  id="horizontal_login_title"
                  placeholder="Title"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="horizontal_login_link"
              title="Link to content"
            >
              Link to content
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <input
                  class="ant-input ant-input-lg"
                  id="horizontal_login_link"
                  placeholder="Link"
                  type="text"
                  value=""
                />
              </div>
            </div>
          </div>
        </div>
        <button
          class="ant-btn ant-btn-link ant-btn-lg"
          type="button"
        >
          <span>
            Creating an event?
          </span>
        </button>
        <div
          class="ant-row ant-form-item"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class=""
              for="horizontal_login_whyShare"
              title="Why are you sharing this?"
            >
              Why are you sharing this?
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <textarea
                  class="ant-input ant-input-lg"
                  id="horizontal_login_whyShare"
                  placeholder="Why are you sharing this?"
                  rows="3"
                  style="font-size: 16px;"
                />
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item ant-form-item-has-success"
        >
          <div
            class="ant-col ant-form-item-label"
          >
            <label
              class="ant-form-item-required"
              for="horizontal_login_circleIds"
              title="Where are you sharing this to?"
            >
              Where are you sharing this to?
            </label>
          </div>
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <div
                  class="ant-select ant-select-lg ant-select-multiple ant-select-show-arrow"
                >
                  <div
                    class="ant-select-selector"
                  >
                    <span
                      class="ant-select-selection-item"
                    >
                      <span
                        class="ant-select-selection-item-content"
                      >
                        The greatest circle
                      </span>
                      <span
                        aria-hidden="true"
                        class="ant-select-selection-item-remove"
                        style="user-select: none;"
                        unselectable="on"
                      >
                        <span
                          aria-label="close"
                          class="anticon anticon-close"
                          role="img"
                        >
                          <svg
                            aria-hidden="true"
                            data-icon="close"
                            fill="currentColor"
                            focusable="false"
                            height="1em"
                            viewBox="64 64 896 896"
                            width="1em"
                          >
                            <path
                              d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                            />
                          </svg>
                        </span>
                      </span>
                    </span>
                    <span
                      class="ant-select-selection-search"
                      style="width: 0px;"
                    >
                      <input
                        aria-activedescendant="horizontal_login_circleIds_list_0"
                        aria-autocomplete="list"
                        aria-controls="horizontal_login_circleIds_list"
                        aria-haspopup="listbox"
                        aria-owns="horizontal_login_circleIds_list"
                        autocomplete="off"
                        class="ant-select-selection-search-input"
                        id="horizontal_login_circleIds"
                        readonly=""
                        role="combobox"
                        style="opacity: 0;"
                        type="search"
                        unselectable="on"
                        value=""
                      />
                      <span
                        aria-hidden="true"
                        class="ant-select-selection-search-mirror"
                      >
                        
                        Â 
                      </span>
                    </span>
                  </div>
                  <span
                    aria-hidden="true"
                    class="ant-select-arrow"
                    style="user-select: none;"
                    unselectable="on"
                  >
                    <span
                      aria-label="down"
                      class="anticon anticon-down ant-select-suffix"
                      role="img"
                    >
                      <svg
                        aria-hidden="true"
                        data-icon="down"
                        fill="currentColor"
                        focusable="false"
                        height="1em"
                        viewBox="64 64 896 896"
                        width="1em"
                      >
                        <path
                          d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                        />
                      </svg>
                    </span>
                  </span>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="ant-row ant-form-item ant-form-item-has-success"
        >
          <div
            class="ant-col ant-form-item-control"
          >
            <div
              class="ant-form-item-control-input"
            >
              <div
                class="ant-form-item-control-input-content"
              >
                <button
                  class="ant-btn ant-btn-primary ant-btn-lg ant-btn-block"
                  data-testid="submitButton"
                  disabled=""
                  type="submit"
                >
                  <span>
                    Submit
                  </span>
                </button>
              </div>
            </div>
          </div>
        </div>
      </form>
    </div>
  </body>,
  "container": <div>
    <form
      class="ant-form ant-form-vertical ant-form-large form"
      id="horizontal_login"
    >
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="horizontal_login_title"
            title="Title of post"
          >
            Title of post
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input ant-input-lg"
                id="horizontal_login_title"
                placeholder="Title"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="horizontal_login_link"
            title="Link to content"
          >
            Link to content
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <input
                class="ant-input ant-input-lg"
                id="horizontal_login_link"
                placeholder="Link"
                type="text"
                value=""
              />
            </div>
          </div>
        </div>
      </div>
      <button
        class="ant-btn ant-btn-link ant-btn-lg"
        type="button"
      >
        <span>
          Creating an event?
        </span>
      </button>
      <div
        class="ant-row ant-form-item"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class=""
            for="horizontal_login_whyShare"
            title="Why are you sharing this?"
          >
            Why are you sharing this?
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <textarea
                class="ant-input ant-input-lg"
                id="horizontal_login_whyShare"
                placeholder="Why are you sharing this?"
                rows="3"
                style="font-size: 16px;"
              />
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item ant-form-item-has-success"
      >
        <div
          class="ant-col ant-form-item-label"
        >
          <label
            class="ant-form-item-required"
            for="horizontal_login_circleIds"
            title="Where are you sharing this to?"
          >
            Where are you sharing this to?
          </label>
        </div>
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <div
                class="ant-select ant-select-lg ant-select-multiple ant-select-show-arrow"
              >
                <div
                  class="ant-select-selector"
                >
                  <span
                    class="ant-select-selection-item"
                  >
                    <span
                      class="ant-select-selection-item-content"
                    >
                      The greatest circle
                    </span>
                    <span
                      aria-hidden="true"
                      class="ant-select-selection-item-remove"
                      style="user-select: none;"
                      unselectable="on"
                    >
                      <span
                        aria-label="close"
                        class="anticon anticon-close"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="close"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M563.8 512l262.5-312.9c4.4-5.2.7-13.1-6.1-13.1h-79.8c-4.7 0-9.2 2.1-12.3 5.7L511.6 449.8 295.1 191.7c-3-3.6-7.5-5.7-12.3-5.7H203c-6.8 0-10.5 7.9-6.1 13.1L459.4 512 196.9 824.9A7.95 7.95 0 00203 838h79.8c4.7 0 9.2-2.1 12.3-5.7l216.5-258.1 216.5 258.1c3 3.6 7.5 5.7 12.3 5.7h79.8c6.8 0 10.5-7.9 6.1-13.1L563.8 512z"
                          />
                        </svg>
                      </span>
                    </span>
                  </span>
                  <span
                    class="ant-select-selection-search"
                    style="width: 0px;"
                  >
                    <input
                      aria-activedescendant="horizontal_login_circleIds_list_0"
                      aria-autocomplete="list"
                      aria-controls="horizontal_login_circleIds_list"
                      aria-haspopup="listbox"
                      aria-owns="horizontal_login_circleIds_list"
                      autocomplete="off"
                      class="ant-select-selection-search-input"
                      id="horizontal_login_circleIds"
                      readonly=""
                      role="combobox"
                      style="opacity: 0;"
                      type="search"
                      unselectable="on"
                      value=""
                    />
                    <span
                      aria-hidden="true"
                      class="ant-select-selection-search-mirror"
                    >
                      
                      Â 
                    </span>
                  </span>
                </div>
                <span
                  aria-hidden="true"
                  class="ant-select-arrow"
                  style="user-select: none;"
                  unselectable="on"
                >
                  <span
                    aria-label="down"
                    class="anticon anticon-down ant-select-suffix"
                    role="img"
                  >
                    <svg
                      aria-hidden="true"
                      data-icon="down"
                      fill="currentColor"
                      focusable="false"
                      height="1em"
                      viewBox="64 64 896 896"
                      width="1em"
                    >
                      <path
                        d="M884 256h-75c-5.1 0-9.9 2.5-12.9 6.6L512 654.2 227.9 262.6c-3-4.1-7.8-6.6-12.9-6.6h-75c-6.5 0-10.3 7.4-6.5 12.7l352.6 486.1c12.8 17.6 39 17.6 51.7 0l352.6-486.1c3.9-5.3.1-12.7-6.4-12.7z"
                      />
                    </svg>
                  </span>
                </span>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ant-row ant-form-item ant-form-item-has-success"
      >
        <div
          class="ant-col ant-form-item-control"
        >
          <div
            class="ant-form-item-control-input"
          >
            <div
              class="ant-form-item-control-input-content"
            >
              <button
                class="ant-btn ant-btn-primary ant-btn-lg ant-btn-block"
                data-testid="submitButton"
                disabled=""
                type="submit"
              >
                <span>
                  Submit
                </span>
              </button>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
